### PR TITLE
Add appointment to MediCLI

### PR DESCRIPTION
**Pull request by ararchch**
Creates functionality to add and delete appointments

### CODE REVIEW

**Comment by Kappaccinoh** [comment]
a lot of checkstyle errors were flagged as part of the CI build, please see the failing cases for (ubuntu-latest) for the fixes - other than that good job with the code @ararchch

File: src/main/java/seedu/address/model/appointment/Appointment.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * Checks if appointment is valid by comparing appointment date against current date.
+     * A valid new appointment can only be in the future, not the past.
+     *
+     * @param appointmentDate Date to check validity of
+     * @return boolean if appointment is valid or not
+     */
+    private boolean isValidAppointment(AppointmentDate appointmentDate) {
```
Comment (officialchengyud) [inline src/main/java/seedu/address/model/appointment/Appointment.java:54]: Would it be more suitable to name the variable 'isValidAppointment' to something like 'isAppointmentDateValid' for future-proofing? E.g In the future when we check whether doctorId or patientId is valid, then 'isValidAppointment' will not just check for date validity.

File: src/main/java/seedu/address/model/appointment/exceptions/AppointmentNotFoundException.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package seedu.address.model.appointment.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified person.
+ */
+public class AppointmentNotFoundException extends RuntimeException {
```
Comment (officialchengyud) [inline src/main/java/seedu/address/model/appointment/exceptions/AppointmentNotFoundException.java:6]: I like how you added custom exception classes throughout the program! I noticed this across the other classes too. Good job Archit.

File: src/test/java/seedu/address/logic/commands/AddAppointmentCommandTest.java
Comment (officialchengyud) [inline src/test/java/seedu/address/logic/commands/AddAppointmentCommandTest.java:1]: The tests look good to me!

File: src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java
Comment (officialchengyud) [inline src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java:1]: This file of code looks clean, nice job :)

**Review by officialchengyud — COMMENTED** [review]
Just some nitpicks. LGTM! Nice job Archit, you've done well.

File: src/main/java/seedu/address/commons/util/IdUtil.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    /**
+     * Updates map with initial values from storage
+     * @throws ExecutionControl.NotImplementedException
+     */
+    public static void initalMapUpdate() throws ExecutionControl.NotImplementedException {
+        throw new ExecutionControl.NotImplementedException("to be implemented");
+    }
+
+}
```
Comment (alfaloo) [inline src/main/java/seedu/address/commons/util/IdUtil.java:110]: clean implementation

File: src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        return toAdd.isSameAppointment(otherAddCommand.toAdd);
+    }
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .add("toAdd", toAdd)
+                .toString();
+    }
+}
```
Comment (alfaloo) [inline src/main/java/seedu/address/logic/commands/AddAppointmentCommand.java:83]: I love you archit, this is goated piece of code

File: src/main/java/seedu/address/logic/parser/AddAppointmentCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
+
+}
```
Comment (alfaloo) [inline src/main/java/seedu/address/logic/parser/AddAppointmentCommandParser.java:54]: its a work of art

File: src/main/java/seedu/address/model/AddressBook.java [hunk 6]
```diff
@@ -101,6 +113,19 @@ public void removePerson(Person key) {
         persons.remove(key);
     }
 
+    public void addAppointment(Appointment appointment) {
```
Comment (alfaloo) [inline src/main/java/seedu/address/model/AddressBook.java:116]: this is rlly nitpicky, but java convention usually has add return a boolean value to indicate if the addition was successful

File: src/main/java/seedu/address/model/AddressBook.java [hunk 7]
```diff
@@ -101,6 +113,19 @@ public void removePerson(Person key) {
...
+        appointments.add(appointment);
+    }
+
+    public void deleteAppointment(Appointment appointment) {
```
Comment (alfaloo) [inline src/main/java/seedu/address/model/AddressBook.java:120]: same for delete, see above

File: src/main/java/seedu/address/model/appointment/UniqueAppointmentList.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.appointment.exceptions.AppointmentNotFoundException;
+import seedu.address.model.appointment.exceptions.DuplicateAppointmentException;
+
+/**
```
Comment (alfaloo) [inline src/main/java/seedu/address/model/appointment/UniqueAppointmentList.java:14]: rlly detailed javadocs, I love it

File: src/main/java/seedu/address/ui/AppointmentCard.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+     * Creates a {@code AppointmentCard} with the given {@code Appointment} and index to display.
+     */
+    public AppointmentCard(Appointment appt, int displayedIndex) {
+        super(FXML);
+        this.appointment = appt;
+        id.setText(displayedIndex + ". ");
+        doctorNric.setText(appointment.getDoctoNric().nric);
+        patientNric.setText(appointment.getPatientNric().nric);
+        appointmentDate.setText(appointment.getAppointmentDate().toString());
+        appointmentId.setText(appointment.getAppointmentId().appointmentId);
```
Comment (alfaloo) [inline src/main/java/seedu/address/ui/AppointmentCard.java:49]: goated fxml

File: src/test/java/seedu/address/logic/commands/AddAppointmentCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    void toString_returnsValidString() {
+        modelManager.addPerson(ALICE);
+        modelManager.addPerson(BROWN);
+        Appointment appt = new Appointment(BROWN.getNric(), ALICE.getNric(), new AppointmentDate("2024-09-01"));
+        AddAppointmentCommand ad = new AddAppointmentCommand(appt);
+        assertEquals(ad.toString(), "seedu.address.logic.commands.AddAppointmentCommand{toAdd="
+                + appt.toString()
+                + "}");
+    }
+}
```
Comment (alfaloo) [inline src/test/java/seedu/address/logic/commands/AddAppointmentCommandTest.java:89]: nice and detailed tests

File: src/test/java/seedu/address/model/appointment/AppointmentDateTest.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    @Test
+    void testHashCode() {
+    }
+
+    @Test
+    void compareTo() {
+    }
+}
```
Comment (alfaloo) [inline src/test/java/seedu/address/model/appointment/AppointmentDateTest.java:23]: are you going to implement these in future iterations?

**Review by alfaloo — APPROVED** [review]

**Pull request merged**